lib/src/revset.rs

```diff
@@ -2150,45 +2151,94 @@ impl PartialSymbolResolver for GitRefResolver {
+            }
+        } else {
+            Ok(None)
+        }
+    }
+}
+
+#[derive(Default)]
+struct ChangePrefixResolver<'a> {
+    context: Option<&'a IdPrefixContext>,
+}
+
+impl PartialSymbolResolver for ChangePrefixResolver<'_> {
+    fn resolve_symbol(
+        &self,
+        repo: &dyn Repo,
+        symbol: &str,
+    ) -> Result<Option<Vec<CommitId>>, RevsetResolutionError> {
+        if let Some(prefix) = to_forward_hex(symbol).as_deref().and_then(HexPrefix::new) {
+            let resolution = self
+                .context
+                .as_ref()
+                .map(|ctx| ctx.resolve_change_prefix(repo, &prefix))
+                .unwrap_or_else(|| repo.resolve_change_id_prefix(&prefix));
+            match resolution {
+                PrefixResolution::AmbiguousMatch => Err(
+                    RevsetResolutionError::AmbiguousChangeIdPrefix(symbol.to_owned()),
+                ),
+                PrefixResolution::SingleMatch(ids) => Ok(Some(ids)),
+                PrefixResolution::NoMatch => Ok(None),
+            }
+        } else {
+            Ok(None)
+        }
+    }
+}
 
 /// Resolves branches, remote branches, tags, git refs, and full and abbreviated
 /// commit and change ids.
 pub struct DefaultSymbolResolver<'a> {
     repo: &'a dyn Repo,
-    commit_id_resolver: PrefixResolver<'a, CommitId>,
-    change_id_resolver: PrefixResolver<'a, Vec<CommitId>>,
+    commit_id_resolver: CommitPrefixResolver<'a>,
+    change_id_resolver: ChangePrefixResolver<'a>,
 }
 
 impl<'a> DefaultSymbolResolver<'a> {
     pub fn new(repo: &'a dyn Repo) -> Self {
         DefaultSymbolResolver {
             repo,
-            commit_id_resolver: Box::new(|repo, prefix| {
-                repo.index().resolve_commit_id_prefix(prefix)
-            }),
-            change_id_resolver: Box::new(|repo, prefix| repo.resolve_change_id_prefix(prefix)),
+            commit_id_resolver: Default::default(),
+            change_id_resolver: Default::default(),
         }
     }
 
-    pub fn with_commit_id_resolver(
-        mut self,
-        commit_id_resolver: PrefixResolver<'a, CommitId>,
-    ) -> Self {
-        self.commit_id_resolver = commit_id_resolver;
-        self
-    }
-
-    pub fn with_change_id_resolver(
-        mut self,
-        change_id_resolver: PrefixResolver<'a, Vec<CommitId>>,
-    ) -> Self {
-        self.change_id_resolver = change_id_resolver;
+    pub fn with_id_prefix_context(mut self, id_prefix_context: &'a IdPrefixContext) -> Self {
+        self.commit_id_resolver.context = Some(id_prefix_context);
+        self.change_id_resolver.context = Some(id_prefix_context);
         self
     }
 
     fn partial_resolvers(&self) -> impl Iterator<Item = &dyn PartialSymbolResolver> {
-        DEFAULT_RESOLVERS.iter().cloned()
+        let mut out = DEFAULT_RESOLVERS.to_vec();
+        out.push(&self.commit_id_resolver);
+        out.push(&self.change_id_resolver);
+        out.into_iter()
```

Comment:
Since the return type is an iterator, we can use `.chain()` or `itertools::chain!()` macro.

```rust
let prefix_resolvers: [&dyn PartialSymbolResolver; 2] = ..; // to coerce type
itertools::chain!(
    DEFAULT_RESOLVERS.iter().copied(),
    prefix_resolvers,
    extensions...
)
```